File: CaCompteKit/Sources/Catalog/Games/SkyjoRulesV1.swift
import Domain

/// Doc 04 « Le cas Skyjo, en détail » — le jeu de référence du projet, celui qui justifie la
/// couche impérative. Seuls `validate` (exclusivité du joueur qui ferme) et `score` (le
/// doublement) sont propres à Skyjo ; `endCheck`/`standings` restent ceux de l'extension par
/// défaut de `GameRules`, parce que `end` et `tieBreak` de `skyjo.json` suffisent à les décrire.
public struct SkyjoRulesV1: GameRules {
  public static let engineID = "skyjo.v1"

  public init() {}

  public func validate(_ draft: RoundDraft, in state: MatchState, definition: GameDefinition)
    -> ValidationResult
  {
    let closers = draft.inputs.filter { $0.modifiers.contains(.closedRound) }
    guard closers.count == 1 else {
      return .invalid([
        ValidationError(field: .modifier(.closedRound), message: "Un seul joueur ferme la manche.")
      ])
    }

    let extremes = draft.inputs.filter { $0.rawValue < -24 || $0.rawValue > 156 }
    return extremes.isEmpty ? .valid : .warning(["Score inhabituel, à vérifier."])
  }

  public func score(_ draft: RoundDraft, in state: MatchState, definition: GameDefinition)
    -> [ScoreEntry]
  {
    let doublingEnabled = state.variants.bool("doublePenalty", default: true)
    let lowest = draft.inputs.map(\.rawValue).min() ?? 0
    let names = Dictionary(uniqueKeysWithValues: state.participants.map { ($0.id, $0.displayName) })

    return draft.inputs.map { input in
      let closed = input.modifiers.contains(.closedRound)
      let isStrictlyLowest =
        input.rawValue == lowest
        && draft.inputs.filter { $0.rawValue == lowest }.count == 1
      let penalised = doublingEnabled && closed && !isStrictlyLowest && input.rawValue > 0

      return ScoreEntry(
        participantID: input.participantID,
        rawValue: input.rawValue,
        computedValue: penalised ? input.rawValue * 2 : input.rawValue,
        explanation: penalised
          ? "Score doublé : \(names[input.participantID] ?? "ce joueur") a fermé la manche sans le score le plus bas."
          : nil,
        detail: nil,
        modifiers: input.modifiers
      )
    }
  }
}

File: android/catalog/src/main/kotlin/com/cacompte/catalog/games/SkyjoRulesV1.kt
package com.cacompte.catalog.games

import com.cacompte.domain.model.MatchState
import com.cacompte.domain.model.ModifierID
import com.cacompte.domain.model.RoundDraft
import com.cacompte.domain.model.ScoreEntry
import com.cacompte.domain.model.ValidationError
import com.cacompte.domain.model.ValidationResult
import com.cacompte.domain.rules.GameDefinition
import com.cacompte.domain.rules.GameRules

/**
 * Miroir de `SkyjoRulesV1.swift` — doc 04 « Le cas Skyjo, en détail ». Seuls `validate`
 * (exclusivité du joueur qui ferme) et `score` (le doublement) sont propres à Skyjo ;
 * `endCheck`/`standings` restent ceux de l'extension par défaut de [GameRules], parce que `end`
 * et `tieBreak` de `skyjo.json` suffisent à les décrire.
 */
class SkyjoRulesV1 : GameRules {
    override val engineID: String = ENGINE_ID

    override fun validate(
        draft: RoundDraft,
        state: MatchState,
        definition: GameDefinition,
    ): ValidationResult {
        val closers = draft.inputs.filter { ModifierID.closedRound in it.modifiers }
        if (closers.size != 1) {
            return ValidationResult.Invalid(
                listOf(
                    ValidationError(
                        ValidationError.Field.ModifierField(ModifierID.closedRound),
                        "Un seul joueur ferme la manche.",
                    ),
                ),
            )
        }

        val extremes = draft.inputs.filter { it.rawValue < -24 || it.rawValue > 156 }
        return if (extremes.isEmpty()) {
            ValidationResult.Valid
        } else {
            ValidationResult.Warning(
                listOf("Score inhabituel, à vérifier."),
            )
        }
    }

    override fun score(
        draft: RoundDraft,
        state: MatchState,
        definition: GameDefinition,
    ): List<ScoreEntry> {
        val doublingEnabled = state.variants.bool("doublePenalty", default = true)
        val lowest = draft.inputs.minOfOrNull { it.rawValue } ?: 0
        val names = state.participants.associate { it.id to it.displayName }

        return draft.inputs.map { input ->
            val closed = ModifierID.closedRound in input.modifiers
            val isStrictlyLowest =
                input.rawValue == lowest && draft.inputs.count { it.rawValue == lowest } == 1
            val penalised = doublingEnabled && closed && !isStrictlyLowest && input.rawValue > 0

            ScoreEntry(
                participantID = input.participantID,
                rawValue = input.rawValue,
                computedValue = if (penalised) input.rawValue * 2 else input.rawValue,
                explanation =
                    if (penalised) {
                        "Score doublé : ${names[input.participantID] ?: "ce joueur"} a fermé la manche sans le score le plus bas."
                    } else {
                        null
                    },
                detail = null,
                modifiers = input.modifiers,
            )
        }
    }

    companion object {
        const val ENGINE_ID = "skyjo.v1"
    }
}
